### PR TITLE
[next] Support Edge API Endpoints

### PR DESCRIPTION
**Pull request by Schniz**
### Related Issues

- #7898
- #7855
- I have merged #7901 by @ijjk into this PR, to ensure this does not break

#### Tests

- [ ] The code changed/added as part of this PR has been covered with tests
- [ ] All tests pass locally with `yarn test-unit`

#### Code Review

- [ ] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/next/src/index.ts [hunk 1]
```diff
@@ -2575,3 +2584,23 @@ export const prepareCache: PrepareCache = async ({
...
+  pagesDir: string;
+  entryPath: string;
+  outputDirectory: string;
+}) {
+  const [pages, middlewareManifest] = await Promise.all([
+    glob('**/!(_middleware).js', params.pagesDir),
+    getMiddlewareManifest(params.entryPath, params.outputDirectory),
+  ]);
+
+  // Edge Functions do not consider as Serverless Functions
```
Comment (cramforce) [inline packages/next/src/index.ts:2599]: Nit: Typo "are not considered"

Comment (Schniz) [reply packages/next/src/index.ts:2599]:
```suggestion
  // Edge Functions are not considered as Serverless Functions
```

File: packages/next/src/index.ts [hunk 2]
```diff
@@ -2575,3 +2584,23 @@ export const prepareCache: PrepareCache = async ({
...
+  const [pages, middlewareManifest] = await Promise.all([
+    glob('**/!(_middleware).js', params.pagesDir),
+    getMiddlewareManifest(params.entryPath, params.outputDirectory),
+  ]);
+
+  // Edge Functions are not considered as Serverless Functions
+  for (const edgeFunctionFile of Object.keys(
+    middlewareManifest?.functions ?? {}
+  )) {
+    delete pages[edgeFunctionFile.slice(1) + '.js'];
```
Comment (ijjk) [inline packages/next/src/index.ts:2612]: Are we planning to add an error when `runtime: 'edge'` is used with a `getStaticProps` page? Currently it isn't compatible with the `Prerender` type as it expects the `lambda` output so will break for that case.

Comment (Schniz) [reply packages/next/src/index.ts:2612]: That's a good question, but right now with https://github.com/vercel/next.js/pull/37481 it's only going to be api endpoints and not pages, so `Prerender` is not expected to be emitted for this output.

cc @shuding that mentioned some work related to pages 👀

Comment (ijjk) [reply packages/next/src/index.ts:2612]: Might be good to add an invariant here to ensure we only delete API pages for now to be safe then

Comment (shuding) [reply packages/next/src/index.ts:2612]: Although we don't have FS in the edge runtime, we still allow `getStaticProps` together with `runtime: 'edge'`. It's just that it currently behaves like SSR.

Comment (shuding) [reply packages/next/src/index.ts:2612]: Soon we will have to remove this constraint as we're refactoring the edge SSR implementation to use functions here :)

**Review by shuding — APPROVED** [review]

**Comment by Schniz** [comment]
@ijjk can you re-review?

**Review by ijjk — APPROVED** [review]

**Pull request merged**